### ISSUE TITLE
Internal CmdStringHolder does not work as a string correctly

### ISSUE THREAD

**Issue by dragon512** [SCons/scons#3428]
Filing the bug as requested on Discord.
Found with Scons 3.0, 3.1 (did not test with older 2.x)
the example shown below is with windows, however, repeats on any OS I tried this on

The example showing this case is simple:

```
env = Environment()
env.Append(LIBPATH=["path1/sub1","path1/sub2"])
lst = env.Flatten(env.subst_list("$LIBPATH"))
print( [type(i) for i in lst])
for i in lst:
    env.Dir(i)
```
 
This was showing up in a less simple set up of subst() when string where turned into Node objects

stack is:
```
scons: Reading SConscript files ...
[<class 'SCons.Subst.CmdStringHolder'>, <class 'SCons.Subst.CmdStringHolder'>]
TypeError: expected string or bytes-like object:
  File "C:\Users\jkenny\code\dev\scons-issues\tmp\SConstruct", line 28:
    env.Dir(i)
  File "c:\users\jkenny\.virtualenvs\scons-issues-v8c7wvel\lib\site-packages\scons\SCons\Environment.py", line 1978:
    s = self.subst(name)
  File "c:\users\jkenny\.virtualenvs\scons-issues-v8c7wvel\lib\site-packages\scons\SCons\Environment.py", line 514:
    return SCons.Subst.scons_subst(string, self, raw, target, source, gvars, lvars, conv)
  File "c:\users\jkenny\.virtualenvs\scons-issues-v8c7wvel\lib\site-packages\scons\SCons\Subst.py", line 591:
    result = remove.sub('', result)
```

**Comment by dragon512** [comment]
Should add the workaround is to do this:
```
for i in lst:
    env.Dir(str(i))
```

**Comment by mwichmann** [comment]
That's pretty odd since that chunk of code is gated by line 580:
```
if is_String(result):
```

**Comment by dragon512** [comment]
so is_String() returns True for this as you point out.

however the class itself I think is missing a function. I am not sure what it is, however it looks like this issue is its use in the regex remove.sub() call

**Comment by mwichmann** [comment]
Indeed, the re module does seem unhappy with this:
```
> /home/mats/github/scons/src/engine/SCons/Subst.py(592)scons_subst()
-> result = remove.sub('', result)
(Pdb) s
TypeError: expected string or bytes-like object
> /home/mats/github/scons/src/engine/SCons/Subst.py(592)scons_subst()
-> result = remove.sub('', result)
(Pdb) type(result)
<class 'SCons.Subst.CmdStringHolder'>
```

I can take this completely out of SCons and show that a `UserString` instance (`CmdStringHolder` subclasses `UserString`) will get `re.sub` to complain:
```
>>> import re, collections
>>> class AString(str):
...     pass
...
>>> a = AString("foobar")
>>> re.sub('a', '', a)
'foobr'
>>> 
>>> b = collections.UserString("barfoo")
>>> re.sub('a', '', b)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/lib64/python3.7/re.py", line 192, in sub
    return _compile(pattern, flags).sub(repl, string, count)
TypeError: expected string or bytes-like object
>>> 
```

**Comment by dragon512** [comment]
How dangerous is it to change CmdStringHolder to use str as a base class vs collections.UserString?

That is the only fix I see, minus just adding a forced str() at the point this called?

**Comment by mwichmann** [comment]
In *theory* all of the places where scons subclasses UserString, UserList, UserDict could be replaced by subclassing the original type - this is now possible and was not when scons was first written.  But I've spot-checked that and found a couple cases where it doesn't just drop-in work (and a couple where it does).  Just tried a drop-in here and it didn't, which doesn't mean the problem couldn't be fixed...
```
TypeError: 'literal' is an invalid keyword argument for str():
...
  File "/home/mats/github/scons/src/script/../engine/SCons/Subst.py", line 827:
    x = CmdStringHolder(x, literal=literal)
```

**Comment by dragon512** [comment]
I can repeat this as well. I don't see how the user can subclass from str and pass extra args on the __init__(). it seems to expect only (object, encoding='utf-8', errors='strict') as valid values

**Comment by bdbaddog** [comment]
Likely a reasonable fix is to change this logic from subst()

```
    if isinstance(strSubst, str) and '$' not in strSubst:
        return strSubst
```
I'm guessing isinstance(cmd_string_instance,str) is false and thus being evealuated and not checked for the absense of $ and if that's true simply getting returned.

**Comment by dragon512** [comment]
So looking at this in more detail.. looking at possible quick fixes

1) do the workaround I am doing which seems to be fine, however, other people may bump into this still and doing this may not be obvious and or hacky
2) change the code that is doing the remove to turn the object in the string. ( might be less hacky as users don't need to do it directly)
3) given you are working on a new set of objects to be stored as the parsed object set for that string, that new code could just remove the object that represents the ($ $) so that the regexp stuff can just be skipped ( as it would be slower than just not processing the items between ValueTypes ESCAPE_START and ESCAPE_END

**Comment by dragon512** [comment]
Adding a note based on discord discussion. Changing the subst() logic to have (around line 413)

```
if (isinstance(strSubst, str) and '$' not in strSubst) or isinstance(strSubst, CmdStringHolder):
        return strSubst
``` 

seems to address the problem as it avoids subst() trying to re-eval the CmdStringHolder ( which should be already eval'ed) This seems to solve this issues

**Comment by mwichmann** [comment]
see also #2339

**Comment by bdbaddog** [comment]
Fixed by #3434
